lib/supavisor/handlers/proxy/client.ex

```diff
@@ -0,0 +1,784 @@
+      upstream_tls_ca: info.tenant.upstream_tls_ca,
+      upstream_verify: info.tenant.upstream_verify
+    }
+
+    %{
+      data
+      | tenant: info.tenant.external_id,
+        user: user,
+        timeout: info.user.pool_checkout_timeout,
+        ps: info.tenant.default_parameter_status,
+        proxy_type: proxy_type,
+        id: id,
+        heartbeat_interval: info.tenant.client_heartbeat_interval * 1000,
+        db_name: db_name,
+        mode: mode,
+        auth: auth
+    }
+  end
+
+  @spec auth_secrets(map, String.t(), term(), non_neg_integer()) ::
+          {:ok, Supavisor.secrets()} | {:error, term()}
+  ## password secrets
+  def auth_secrets(%{user: user, tenant: %{require_user: true}}, _, _, _) do
+    secrets = %{db_user: user.db_user, password: user.db_password, alias: user.db_user_alias}
+    {:ok, {:password, fn -> secrets end}}
+  end
+
+  ## auth_query secrets
+  def auth_secrets(info, db_user, key, ttl) do
+    fetch = fn _key ->
+      case get_secrets(info, db_user) do
+        {:ok, _} = resp -> {:commit, {:cached, resp}, ttl: ttl}
+        {:error, _} = resp -> {:ignore, resp}
+      end
+    end
+
+    case Cachex.fetch(Supavisor.Cache, key, fetch) do
+      {:ok, {:cached, value}} -> value
+      {:commit, {:cached, value}, _opts} -> value
+      {:ignore, resp} -> resp
+    end
+  end
+
+  @spec get_secrets(map, String.t()) :: {:ok, {:auth_query, fun()}} | {:error, term()}
+  def get_secrets(%{user: user, tenant: tenant}, db_user) do
+    ssl_opts =
+      if tenant.upstream_ssl and tenant.upstream_verify == :peer do
+        [
+          {:verify, :verify_peer},
+          {:cacerts, [Helpers.upstream_cert(tenant.upstream_tls_ca)]},
+          {:server_name_indication, String.to_charlist(tenant.db_host)},
+          {:customize_hostname_check, [{:match_fun, fn _, _ -> true end}]}
+        ]
+      end
+
+    {:ok, conn} =
+      Postgrex.start_link(
+        hostname: tenant.db_host,
+        port: tenant.db_port,
+        database: tenant.db_database,
+        password: user.db_password,
+        username: user.db_user,
+        parameters: [application_name: "Supavisor auth_query"],
+        ssl: tenant.upstream_ssl,
+        socket_options: [
+          Helpers.ip_version(tenant.ip_version, tenant.db_host)
+        ],
+        queue_target: 1_000,
+        queue_interval: 5_000,
+        ssl_opts: ssl_opts || []
+      )
+
+    # kill the postgrex connection if the current process exits unexpectedly
+    Process.link(conn)
+
+    Logger.debug(
+      "ProxyClient: Connected to db #{tenant.db_host} #{tenant.db_port} #{tenant.db_database} #{user.db_user}"
+    )
+
+    resp =
+      with {:ok, secret} <- Helpers.get_user_secret(conn, tenant.auth_query, db_user) do
+        t = if secret.digest == :md5, do: :auth_query_md5, else: :auth_query
+        {:ok, {t, fn -> Map.put(secret, :alias, user.db_user_alias) end}}
+      end
+
+    GenServer.stop(conn, :normal, 5_000)
```

Comment:
Could there be a bug here? Once you stop the process, it will deliver a `{:EXIT, pid, :normal}` message to this process, since you are trapping exits. I don't see such messages being discarded but I may have missed where it happens, so I thought about dropping a comment.

Similarly, if whatever you check out from poolboy crashes, you may also get an EXIT message, although this case can only happen if the pool terminates, so I assume something else is already going wrong.

Reply:
This code occurs [inside Cachex's fetch](https://github.com/supabase/supavisor/blob/6e54020d307caee55f16d4ef833ed03fa9aa3cc9/lib/supavisor/handlers/proxy/client.ex#L613-L626), so it's [linked to an internally spawned process](https://github.com/supabase/supavisor/blob/6e54020d307caee55f16d4ef833ed03fa9aa3cc9/lib/supavisor/handlers/proxy/client.ex#L657-L658)